### PR TITLE
Stop loss open multiply flow

### PR DESCRIPTION
**Pull request by piekczyk**
# [Stop loss open multiply flow](https://app.shortcut.com/oazo-apps/story/4151/stop-loss-form-integration-in-multiply-vault) 
(and all related stories except for unit tests)

<please insert a shortcut link above>
  
## Changes 👷‍♀️
  <Please add short list of changes>
- added stop loss to multiply open flow
  
## How to test 🧪
  <Please explain how to test your changes>
- use NewComponents and StopLossOpenFlow feature toggle and try to open multiply vault with and without stop loss (when added during vault open flow, SL should be enabled in manage vault protection tab)

### CODE REVIEW

File: blockchain/config.ts [hunk 1]
```diff
@@ -353,7 +353,7 @@ const hardhat: NetworkConfig = {
   label: 'Hardhat',
   infuraUrl: `http://localhost:8545`,
   infuraUrlWS: `ws://localhost:8545`,
-  cacheApi: 'https://oazo-bcache-mainnet-staging.new.oasis.app/api/v1',
+  cacheApi: 'http://localhost:3001/v1',
```
Comment (George-field) [inline blockchain/config.ts:356]: Is this correct? shouldn't it be changed back?

File: features/automation/protection/openFlow/openVaultStopLoss.ts [hunk 2]
```diff
@@ -44,11 +46,22 @@ export function applyOpenVaultStopLoss(
...
+  afterCollateralizationRatio: BigNumber
+  afterLiquidationPrice: BigNumber
+  setStopLossCloseType: (type: CloseVaultTo) => void
+  setStopLossLevel: (level: BigNumber) => void
+  stopLossCloseType: CloseVaultTo
+  stopLossLevel: BigNumber
+  depositAmount?: BigNumber
+  generateAmount?: BigNumber
+  afterOutstandingDebt?: BigNumber
+}) {
```
Comment (George-field) [inline features/automation/protection/openFlow/openVaultStopLoss.ts:64]: General question, are we using similar interfaces to this for other components in other places? Also, another thought here is are we following a pattern with the way we define interfaces for components as I see that it is defined in different ways all over the place.

File: features/automation/protection/openFlow/stopLossOpenFlowTransaction.ts [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+import { Vault } from 'blockchain/vaults'
+import { TxHelpers } from 'components/AppContext'
+import { prepareAddTriggerData } from 'features/automation/protection/controls/AdjustSlFormControl'
+import { CloseVaultTo } from 'features/multiply/manage/pipes/manageMultiplyVault'
+import { AddStopLossChange } from 'features/shared/transactions'
+import { transactionToX } from 'helpers/form'
+import { of } from 'rxjs'
+
+export function applyStopLossOpenFlowTransaction<S>(state: S, change: AddStopLossChange) {
+  if (change.kind === 'stopLossTxWaitingForConfirmation') {
```
Comment (George-field) [inline features/automation/protection/openFlow/stopLossOpenFlowTransaction.ts:16]: Love a good if statement 😉, not sure what the preference is, some people may say use a switch here, personally I don't think it is overly important

File: features/borrow/open/pipes/openVaultConditions.ts [hunk 4]
```diff
@@ -56,27 +56,24 @@ export function applyOpenVaultStageCategorisation(state: OpenVaultState) {
...
-      (checkIfStopLossAndGenerate && state.totalSteps === 2)
-    ? 3
-    : !checkIfStopLossAndGenerate && state.totalSteps === 3
-    ? 2
-    : proxyAddress && hasAllowance && state.totalSteps === 4
-    ? 2
-    : state.totalSteps
+  const withStopLossStep = !!(withStopLossStage && generateAmount?.gt(zero))
+
+  const totalSteps = getTotalStepsForOpenVaultFlow({
```
Comment (George-field) [inline features/borrow/open/pipes/openVaultConditions.ts:68]: Nice! Looks a lot cleaner :)

File: features/multiply/open/containers/OpenMultiplyVaultChangesInformation.tsx [hunk 5]
```diff
@@ -163,6 +173,30 @@ export function OpenMultiplyVaultChangesInformation(props: OpenMultiplyVaultStat
           <VaultChangesInformationEstimatedGasFee {...props} />
         </Grid>
       )}
+      {stopLossOpenFlowEnabled && stopLossLevel.gt(zero) && !stopLossSkipped && (
```
Comment (George-field) [inline features/multiply/open/containers/OpenMultiplyVaultChangesInformation.tsx:176]: I like this but we are using a lot of code reuse here, I have created a new component for this that can be used in the future, see https://github.com/OasisDEX/oasis-borrow/pull/1102 for future reference

File: helpers/totalSteps.ts [hunk 6]
```diff
@@ -0,0 +1,49 @@
+export function getTotalStepsForOpenVaultFlow({
```
Comment (George-field) [inline helpers/totalSteps.ts:1]: Nice function, very neat and useful!

**Review by George-field — COMMENTED** [review]
Really nice addition, just a couple of points mentioned, nothing overly major just a few general queries

File: blockchain/config.ts (hunk 1 above)
Comment (piekczyk) [reply blockchain/config.ts:356]: yep, its fine

**Review by piotrkonowrocki — APPROVED** [review]

**Review by piotrkonowrocki — APPROVED** [review]
I'm approving PR because overall code quality is good, but maybe it would be worth it to take a look at the amount of duplication? Is is possible to extract some shared stuff between open borrow/multiply?

**Pull request merged**